### ISSUE TITLE
Overflow/Underflow of pixel value

### ISSUE THREAD

**Issue by kjy09** [ando-khachatryan/HiDDeN#29]
Hi, thanks for your contribution to the source code. I encounter a problem when training the encoder and decoder. After embedding a watermark into the cover image, the pixel value may overflow/underflow, which degrades the visual quality of the encoded image (some weird noises appear). To tackle this problem, I clip the pixel value to ensure it is in the range [-1,1]. However, this leads to a decrease in the bit error rate between the extracted message and the original one. What should I do? Besides, it seems that converting a float value to an integer may also decrease the bit error rate due to the loss of precision. Is there any good solution?

**Comment by ando-khachatryan** [comment]
The original paper computes MSE on watermarked image vs the cover, this discourages pushing the values far from the cover (so >1 and < -1 would be discouraged). This is not a hard constraint though, and the as you point out the values can go outside the range. 
You can torch.clip the pixel values as part of the end-to-end training. Another alternative which I was used was to "quantize" the output, i.e. essentially do a transformation of the real-valued pixed values from [a, b] into integers [0, 255] and back to [-1, 1] or [0, 1]. This not only ensures that the pixel values are within the normal range, it also forces the network NOT to rely on 32-bit precision, simply because that will be converted into 8-bit when the image is saved. 

It is a noise layer, like the Resize or Crop, something like this:

```
import numpy as np
import torch
import torch.nn as nn


def transform(tensor: torch.Tensor, target_range):
    source_min = tensor.min()
    source_max = tensor.max()
    # normalize to [0, 1]
    tensor_target = (tensor - source_min)/(source_max - source_min)
    # move to target range
    tensor_target = tensor_target * (target_range[1] - target_range[0]) + target_range[0]
    return tensor_target


class FourierRound(nn.Module):
    def __init__(self, fourier_coeff_count: int=10):
        super(FourierRound, self).__init__()

        weights = torch.tensor([np.power(-1, n)/(np.pi * n) for n in range(1, fourier_coeff_count+1)])
        scales = torch.tensor([2 * np.pi * n for n in range(1, fourier_coeff_count+1)])
        for _ in range(4):
            weights.unsqueeze_(-1)
            scales.unsqueeze_(-1)

        self.register_buffer('weights', weights)
        self.register_buffer('scales', scales)

    def forward(self, tensor):
        z = torch.mul(self.weights, torch.sin(torch.mul(tensor, self.scales)))
        z = torch.sum(z, dim=0)
        return tensor + z


class Quantization(nn.Module):
    def __init__(self, fourier_coeff_count: int=10, target_range=(0, 255)):
        super(Quantization, self).__init__()
        self.fourier_round = FourierRound(fourier_coeff_count)
        self.target_range = target_range

    def forward(self, noised_and_cover):
        noised_image = noised_and_cover[0]
        original_range = (noised_image.min(), noised_image.max())
        z = transform(noised_image, self.target_range)
        z = self.fourier_round(z)
        z = transform(z, original_range)
        return z

    def __repr__(self):
        return 'quant()'
 
```

**Comment by kjy09** [comment]
Thanks for your helpful suggestions and for providing suitable solutions. This problem has puzzled me a lot.
